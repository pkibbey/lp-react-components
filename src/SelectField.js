import React from 'react'
import { Box } from 'rebass'
import { Select, Label } from '@rebass/forms'
import ErrorText from './ErrorText'

const SelectField = ({
  name,
  value,
  defaultValue,
  updateUserDetail,
  error,
  updateError,
  countries
}) => {
  const getVariant = () => {
    if (error.isError && value === '') {
      return 'textInputErrorDisabled'
    }
    if (error.isError) {
      return 'textInputError'
    }
    if (value === '') {
      return 'textInputDisabled'
    }
    return 'textInput'
  }

  return (
    <Box>
      <Box
        sx={{
          position: 'relative',
          display: 'grid',
          gridTemplateColumns: ['auto', '5fr 4fr']
        }}
      >
        <Label htmlFor={`select-field-${name}`} style={{ display: 'none' }}>
          {name}
        </Label>
        <Select
          data-testid={`select-field-${name}`}
          autoComplete='off'
          variant={getVariant()}
          defaultValue={defaultValue}
          onChange={(event) =>
            updateUserDetail && updateUserDetail(name, event.target.value)
          }
          onBlur={() =>
            updateError && updateError(name, { hasInteracted: true })
          }
          name='region'
          mb={error.isError ? 2 : 4}
          sx={{
            '&:focus, &:hover': {
              outlineColor: 'navyGray',
              outlineWidth: 2,
              outlineStyle: 'auto'
            }
          }}
        >
          {countries.map((option) => (
            <option
              key={option.name}
              value={option.value || 'z1'}
              disabled={option.disabled}
            >
              {option.name}
            </option>
          ))}
        </Select>
      </Box>
      {error && <ErrorText name={name} error={error} />}
    </Box>
  )
}

export default SelectField
