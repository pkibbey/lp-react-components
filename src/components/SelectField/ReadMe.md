Default select field:

```js
<SelectField
  value=""
  options={[
    { label: 'test 1', value: 'test1' },
    { label: 'test 2', value: 'test2' }
  ]}
/>
```

Errored select field:

```js
<SelectField
  value=""
  options={[
    { label: 'test 1', value: 'test1' },
    { label: 'test 2', value: 'test2' }
  ]}
  error={{
    isError: true,
    hasInteracted: true,
    message: 'invalid option'
  }}
/>
```
