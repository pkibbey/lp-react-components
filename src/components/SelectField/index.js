import React from 'react'
import { Box } from 'rebass'
import { Select, Label } from '@rebass/forms'
import ErrorText from '../ErrorText'
import PropTypes from 'prop-types'
import ThemeWrapper from '../ThemeWrapper'

/**
 * Used for selecting a single option from a defined set
 *
 * @visibleName Select Field
 */
const SelectField = ({
  name,
  value,
  defaultValue,
  handleChange,
  error,
  handleBlur,
  options
}) => {
  const isErrored = error && error.isError

  const getVariant = () => {
    if (isErrored && value === '') {
      return 'textInputErrorDisabled'
    }
    if (isErrored) {
      return 'textInputError'
    }
    if (value === '') {
      return 'textInputDisabled'
    }
    return 'textInput'
  }

  return (
    <ThemeWrapper>
      <Box>
        <Box
          sx={{
            position: 'relative',
            display: 'grid',
            gridTemplateColumns: ['auto', '0 5fr 4fr']
          }}
        >
          <Label
            htmlFor={`select-field-${name}`}
            style={{
              // label required for a11y - hide in the UI
              opacity: 0,
              height: 0,
              width: 0,
              pointerEvents: 'none'
            }}
          >
            {name}
          </Label>
          <Select
            data-testid={`select-field-${name}`}
            autoComplete='off'
            defaultValue={defaultValue}
            onChange={(event) =>
              handleChange && handleChange(name, event.target.value)
            }
            onBlur={() =>
              handleBlur && handleBlur(name, { hasInteracted: true })
            }
            name='region'
            mb={isErrored ? 2 : 4}
            py={0}
            px={3}
            variant={getVariant()}
          >
            {options.map((option) => (
              <option
                key={option.name}
                value={option.value}
                disabled={option.disabled}
              >
                {option.name}
              </option>
            ))}
          </Select>
        </Box>
        <ErrorText error={error} mb={3} />
      </Box>
    </ThemeWrapper>
  )
}

SelectField.propTypes = {
  /** The value for the select field */
  value: PropTypes.string.isRequired,
  /** An array of options for the select field  */
  options: PropTypes.array,
  /** Select field label for accessibility */
  name: PropTypes.string,
  /** A callback to fire when the select field changes */
  handleChange: PropTypes.func,
  /** A callback to fire when the select field loses focus */
  handleBlur: PropTypes.func,
  /** A defaultValue for the select field */
  defaultValue: PropTypes.string,
  /** An object describing the error in the select field */
  error: PropTypes.shape({
    isError: PropTypes.bool,
    hasInteracted: PropTypes.bool,
    message: PropTypes.string
  })
}

SelectField.defaultProps = {
  name: 'select-field',
  defaultValue: '',
  options: [],
  error: {}
}

export default SelectField
