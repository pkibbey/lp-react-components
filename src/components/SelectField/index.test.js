import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import SelectField from './'

const OPTIONS = [
  { name: 'Country 1', value: 'country-1' },
  { name: 'Country 2', value: 'country-2' },
  { name: 'Country 3', value: 'country-3' }
]
const mockBlur = jest.fn()
const mockChange = jest.fn()
const SELECT_NAME = 'select1'
const ERROR_MESSAGE = 'error message'

beforeEach(() => {
  jest.clearAllMocks()
})

it('renders a select field', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue={OPTIONS[1].value}
      handleChange={mockChange}
      handleBlur={mockBlur}
      options={OPTIONS}
      isFullWidth
    />
  )
  const combobox = screen.getByRole('combobox')
  const options = screen.getAllByRole('option')
  const error = screen.queryByText(ERROR_MESSAGE)

  expect(error).toBeNull()
  expect(combobox).toHaveAttribute('name', SELECT_NAME)
  expect(combobox).toHaveValue(OPTIONS[1].value)
  expect(options).toHaveLength(3)
  expect(mockChange).not.toHaveBeenCalled()
  expect(mockBlur).not.toHaveBeenCalled()
})

it('renders a select field with an error', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue={OPTIONS[2].value}
      handleChange={mockChange}
      handleBlur={mockBlur}
      error={{ isError: true, message: ERROR_MESSAGE, hasInteracted: true }}
      options={OPTIONS}
    />
  )
  const combobox = screen.getByRole('combobox')
  const options = screen.getAllByRole('option')
  const error = screen.getByText(ERROR_MESSAGE)

  expect(error).toHaveStyle({ color: 'rgb(223, 42, 49)' })
  expect(combobox).toHaveAttribute('name', SELECT_NAME)
  expect(options).toHaveLength(3)
  expect(combobox).toHaveValue(OPTIONS[2].value)
  expect(mockChange).not.toHaveBeenCalled()
  expect(mockBlur).not.toHaveBeenCalled()
})

it('renders a select field with no value', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue={OPTIONS[2].value}
      handleChange={mockChange}
      handleBlur={mockBlur}
      options={OPTIONS}
    />
  )
  const combobox = screen.getByRole('combobox')
  const options = screen.getAllByRole('option')
  const error = screen.queryByText(ERROR_MESSAGE)

  expect(error).toBeNull()
  expect(combobox).toHaveAttribute('name', SELECT_NAME)
  expect(combobox).toHaveValue(OPTIONS[2].value)
  expect(options).toHaveLength(3)
  expect(mockChange).not.toHaveBeenCalled()
  expect(mockBlur).not.toHaveBeenCalled()
})

it('renders a select field with an error and no value', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue=''
      handleChange={mockChange}
      handleBlur={mockBlur}
      error={{ isError: true }}
      options={OPTIONS}
    />
  )
  const combobox = screen.getByRole('combobox')
  const options = screen.getAllByRole('option')
  const error = screen.queryByText(ERROR_MESSAGE)

  expect(error).toBeNull()
  expect(combobox).toHaveAttribute('name', SELECT_NAME)
  expect(combobox).toHaveValue(OPTIONS[0].value)
  expect(options).toHaveLength(3)
  expect(mockChange).not.toHaveBeenCalled()
  expect(mockBlur).not.toHaveBeenCalled()
})

test('select field change event is fired', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue=''
      handleChange={mockChange}
      handleBlur={mockBlur}
      options={OPTIONS}
    />
  )
  const combobox = screen.getByRole('combobox')

  fireEvent.change(combobox, {
    target: { value: OPTIONS[2].value }
  })

  expect(combobox).toHaveValue(OPTIONS[2].value)
  expect(mockChange).toHaveBeenCalledTimes(1)
  expect(mockBlur).not.toHaveBeenCalled()
})

test('select field blur event is fired', () => {
  render(
    <SelectField
      name={SELECT_NAME}
      defaultValue={OPTIONS[0].value}
      handleChange={mockChange}
      handleBlur={mockBlur}
      options={OPTIONS}
    />
  )
  const combobox = screen.getByRole('combobox')

  fireEvent.blur(combobox)

  expect(combobox).toHaveValue(OPTIONS[0].value)
  expect(mockChange).toHaveBeenCalledTimes(0)
  expect(mockBlur).toHaveBeenCalledTimes(1)
})
