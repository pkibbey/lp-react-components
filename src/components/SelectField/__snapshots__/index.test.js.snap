// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a select field 1`] = `
<div
  className="css-vurnku"
>
  <div
    className="css-1765pcw"
  >
    <div
      className="css-15s5z5b"
    >
      <select
        autoComplete="off"
        className="css-19ju6oh"
        data-testid="select-field-select1"
        defaultValue=""
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="choose-country"
        >
          Select country
        </option>
        <option
          value="country-1"
        >
          Country 1
        </option>
        <option
          value="country-2"
        >
          Country 2
        </option>
      </select>
      <svg
        className="css-6ibwe0"
        fill="currentcolor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      className="css-1r4rh9s"
    />
  </div>
</div>
`;

exports[`renders a select field with an error 1`] = `
<div
  className="css-vurnku"
>
  <div
    className="css-16etn8k"
  >
    <div
      className="css-11gqdzc"
    >
      <select
        autoComplete="off"
        className="css-61exfo"
        data-testid="select-field-select2"
        defaultValue=""
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="choose-country"
        >
          Select country
        </option>
        <option
          value="country-1"
        >
          Country 1
        </option>
        <option
          value="country-2"
        >
          Country 2
        </option>
      </select>
      <svg
        className="css-6ibwe0"
        fill="currentcolor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      className="css-1r4rh9s"
    />
  </div>
</div>
`;

exports[`renders a select field with an error and no value 1`] = `
<div
  className="css-vurnku"
>
  <div
    className="css-16etn8k"
  >
    <div
      className="css-11gqdzc"
    >
      <select
        autoComplete="off"
        className="css-77lhji"
        data-testid="select-field-select2"
        defaultValue=""
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="choose-country"
        >
          Select country
        </option>
        <option
          value="country-1"
        >
          Country 1
        </option>
        <option
          value="country-2"
        >
          Country 2
        </option>
      </select>
      <svg
        className="css-6ibwe0"
        fill="currentcolor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      className="css-1r4rh9s"
    />
  </div>
</div>
`;

exports[`renders a select field with no value 1`] = `
<div
  className="css-vurnku"
>
  <div
    className="css-16etn8k"
  >
    <div
      className="css-15s5z5b"
    >
      <select
        autoComplete="off"
        className="css-13563nv"
        data-testid="select-field-select2"
        defaultValue=""
        name="region"
        onBlur={[Function]}
        onChange={[Function]}
      >
        <option
          disabled={true}
          value="choose-country"
        >
          Select country
        </option>
        <option
          value="country-1"
        >
          Country 1
        </option>
        <option
          value="country-2"
        >
          Country 2
        </option>
      </select>
      <svg
        className="css-6ibwe0"
        fill="currentcolor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      className="css-1r4rh9s"
    />
  </div>
</div>
`;
